app/lib/actions/katello/host/hypervisors_update.rb

```diff
@@ -13,37 +13,67 @@ def run
+        def load_resources
+          @organizations = ::Organization.where(label: hypervisors_field(:organization_label)).map { |org| [org.label, org] }.to_h
+          hosts_by_uuid = ::Host.eager_load(:subscription_facet).where(katello_subscription_facets: { uuid: hypervisors_field(:uuid) })
+          @hosts = hosts_by_uuid.map { |host| [host.subscription_facet.uuid, host] }.to_h
+
+          duplicate_names = {}
+          duplicate_name_orgs = {}
+          @hypervisors.each do |hypervisor|
+            next if @hosts.key?(hypervisor[:uuid])
 
-          # Since host names must be unique yet hypervisors may have unique subscription
-          # facets in different orgs
-          sanitized_name = ::Katello::Host::SubscriptionFacet.sanitize_name(hypervisor_json[:name])
-          duplicate_name = "virt-who-#{sanitized_name}-#{organization.id}"
-          host = ::Katello::Host::SubscriptionFacet.find_by(:uuid => hypervisor_json[:uuid]).try(:host)
-          host ||= ::Host.find_by(:name => duplicate_name)
-          if host && host.organization.try(:id) != organization.id
-            fail _("Host '%{name}' does not belong to an organization") % {:name => host.name} unless host.organization
-            host = nil
+            duplicate_name, org = duplicate_name(hypervisor)
+            duplicate_names[duplicate_name] = hypervisor[:uuid]
+            duplicate_name_orgs[duplicate_name] = org
           end
 
-          host ||= create_host_for_hypervisor(duplicate_name, organization)
-          host.subscription_facet ||= ::Katello::Host::SubscriptionFacet.new
-          host.subscription_facet.host_id = host.id
-          host.subscription_facet.uuid = hypervisor_json[:uuid]
-          host.subscription_facet.import_database_attributes(host.subscription_facet.candlepin_consumer.consumer_attributes)
-          host.subscription_facet.save!
-          host.subscription_facet.update_subscription_status
-          host.save!
+          hosts_by_dup_name = ::Host.preload(:subscription_facet).where(name: duplicate_names.keys)
+          @hosts.merge!(hosts_by_dup_name.map { |host| [duplicate_names[host.name], host] }.to_h)
+
+          @hosts.values.each do |host|
+            if host.organization_id.nil? || host.organization_id != duplicate_name_orgs[host.name].try(:id)
+              fail _("Host '%{name}' does not belong to an organization") % {:name => host.name} unless host.organization
+            end
+          end
+
+          # remaining hypervisors
+          @hypervisors.each do |hypervisor|
+            next if @hosts.key?(hypervisor[:uuid])
+            duplicate_name, org = duplicate_name(hypervisor)
+            @hosts[hypervisor[:uuid]] = create_host_for_hypervisor(duplicate_name, org)
+          end
+          candlepin_data = ::Katello::Resources::Candlepin::Consumer.get(uuid: @hosts.keys)
+          @candlepin_attributes = candlepin_data.map { |consumer| [consumer[:uuid], consumer] }.to_h
+        end
```

Comment:
still reiviewing this, i tested it and it seemed fine, but i'm still trying to grok what all this is doing.  Can you think of easy ways to split htis method up to make it a bit easier to read?